### PR TITLE
new: Support TLS-PSK (TLS 1.3)

### LINKED ISSUE #107: feature: extensionPreSharedKey support
Hey,
is there a way in Utls to add PSK/pre_shared_key/extensionPreSharedKey,
all i can find in utls is that somewhere you guys say: TODO: wait for someone who cares about PSK to implement
Thanks.

### LINKED ISSUE #182: The session resumption doesn`t work
I tried using HelloChrome_100_PSK to see the session resumption

1. first request

```
	config := tls.Config{
		ClientSessionCache: tls.NewLRUClientSessionCache(32),
		ServerName:         requestHostname,
	}
	dialConn, err := net.DialTimeout("tcp", addr, dialTimeout)

	uTlsConn := tls.UClient(dialConn, config, tls.HelloChrome_100)
	uTlsConn.Handshake()
```
2. second request to see if session resumption is possible

```
		cacheKey := config.ServerName
		session, ok := config.ClientSessionCache.Get(cacheKey)
		if !ok || session == nil {
			return
		}
		sessionTicket := session.SessionTicket()

		ticketAge := uint32(time.Now().Sub(session.ReciveAt()) / time.Millisecond)
		obfuscatedTicketAge := ticketAge + session.AgeAdd()


	uTlsConn := tls.UClient(dialConn, config, tls.HelloCustom)
	preset, err := tls.UTLSIdToSpec(tls.HelloChrome_100_PSK) // correct
	if err != nil {
		return "", err
	}
	if pskExt, ok := preset.Extensions[len(preset.Extensions)-1].(*tls.FakePreSharedKeyExtension); ok {
		pskExt.PskIdentities = []tls.PskIdentity{ // must set identity
			{
				Label:               sessionTicket, // change this
				ObfuscatedTicketAge: obfuscatedTicketAge,          // change this
			},
		}
		// each fake binder is 32 bytes of zeros
		pskExt.PskBinders = [][]byte{ // must set psk binders
			{
				0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // change this
			},
		} // byte slices
	}
	uTlsConn.ApplyPreset(&preset) // make sure to apply preset to the connection

uTlsConn.Handshake()
```
error occurred:
**#> HttpGetCustom() failed: uTlsConn.Handshake() error: local error: tls: bad record MAC**

image

**Is PSK binders is wrong or some other reason?**

deep into the code
handshake_client.go -> loadSession

```
	ticketAge := uint32(c.config.time().Sub(session.receivedAt) / time.Millisecond)
	identity := pskIdentity{
		label:               session.sessionTicket,
		obfuscatedTicketAge: ticketAge + session.ageAdd,
	}
	hello.pskIdentities = []pskIdentity{identity}
	hello.pskBinders = [][]byte{make([]byte, cipherSuite.hash.Size())}

	// Compute the PSK binders. See RFC 8446, Section 4.2.11.2.
	psk := cipherSuite.expandLabel(session.masterSecret, "resumption",
		session.nonce, cipherSuite.hash.Size())
	earlySecret = cipherSuite.extract(psk, nil)
	binderKey = cipherSuite.deriveSecret(earlySecret, resumptionBinderLabel, nil)
	transcript := cipherSuite.hash.New()
	helloBytes, err := hello.marshalWithoutBinders()
```

It seems like the PSK binder calculation  is correct, so I cannot find where the issue is

### PR DESCRIPTION
**Pull request by gaukas**
# Note

Implement `UtlsPreSharedKeyExtension` to enable REAL PSK resumption.
Redesign `FakePreSharedKeyExtension` and `HardcodedPreSharedKeyExtension` to mimic PSK extension.

Close #107.
Close #182.
Related to #63. 

This PR is based off branch:[x25519kyber768](https://github.com/refraction-networking/utls/tree/x25519kyber768). 
Close #223 for duplicated commits.

# Commit

In `u_pre_shared_key.go`, create `PreSharedKeyExtension` as an interface, with ~3~2 implementations:
- `UtlsPreSharedKeyExtension` implements full support for `pre_shared_key` unless resuming after seeing HRR (which breaks the connection).
- `FakePreSharedKeyExtension` allows user to hardcode Identities and Binders to be sent in the extension without setting the internal states. Therefore if the server accepts the PSK and tries to resume, the connection fails.

### CODE REVIEW

File: u_pre_shared_key.go [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+//
+// It does not compute binders based on ClientHello, but uses the binders specified instead.
+//
+// TODO: Only one of FakePreSharedKeyExtension and HardcodedPreSharedKeyExtension should
+// be kept, the other one should be just removed. We still need to learn more of the safety
+// of hardcoding both Identities and Binders without recalculating the latter.
+type HardcodedPreSharedKeyExtension struct {
+	Identities []PskIdentity `json:"identities"`
+	Binders    [][]byte      `json:"binders"`
+}
```
Comment (gaukas) [inline u_pre_shared_key.go:335]: `HardcodePreSharedKeyExtension` should embed `UnimplementedPreSharedKeyExtension`. I will push a change. 
```suggestion
type HardcodedPreSharedKeyExtension struct {
        UnimplementedPreSharedKeyExtension

	Identities []PskIdentity `json:"identities"`
	Binders    [][]byte      `json:"binders"`
}
```

File: u_pre_shared_key.go [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+// of hardcoding both Identities and Binders without recalculating the latter.
+type FakePreSharedKeyExtension struct {
+	UnimplementedPreSharedKeyExtension
+
+	CipherSuite   uint16 `json:"cipher_suite"`   // this is used to compute the binder
+	SessionSecret []byte `json:"session_secret"` // this is used to compute the binder
+
+	Identities []PskIdentity `json:"identities"`
+	binders    [][]byte
+}
```
Comment (gaukas) [inline u_pre_shared_key.go:207]: We will remove the current implementation of `FakePreSharedKeyExtension` and use `HardcodedPreSharedKeyExtension` instead. The final naming is subjected to change. 

See https://github.com/refraction-networking/utls/issues/107#issuecomment-1679930782 for the discussion supporting this change.

File: examples/tls-psk/main.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	tlsConnPSK := tls.UClient(tcpConnPSK, &tls.Config{
+		ServerName: strings.Split(serverAddr, ":")[0],
+		// ClientSessionCache: csc, // set this will cause PSK to fail. This is a bug...
+	}, tls.HelloChrome_100_PSK, &tls.UtlsPreSharedKeyExtension{
+		ClientSessionCacheOverride: csc, // ONLY set your own ClientSessionCache here if you want to use PSK
+	})
```
[the remark this answers is not in the document]
Comment (zeeker999) [reply examples/tls-psk/main.go:93]: Skip the session resumption handling here works for me, not sure this is the correct fix.
```diff
--- a/u_parrots.go
+++ b/u_parrots.go
@@ -2454,6 +2454,15 @@ func (uconn *UConn) ApplyPreset(p *ClientHelloSpec) error {
                                        session = cs.session
                                }
                                // TLS 1.3 (PSK) resumption is handled by PreSharedKeyExtension in MarshalClientHello()
+                               if session != nil && session.version == VersionTLS13 {
+                                       for _, ext := range uconn.Extensions {
+                                               _, ok := ext.(*UtlsPreSharedKeyExtension)
+                                               if !ok {
+                                                       continue
+                                               }
+                                               cs = nil
+                                       }
+                               }
                        }
                        err := uconn.SetSessionState(cs)
                        if err != nil {
```

[the remark this answers is not in the document]
Comment (gaukas) [reply examples/tls-psk/main.go:93]: Thanks for pointing that out. It actually makes sense... If `Config` have no `ClientSessionCache`, then `SessionTicketExtension` wouldn't be able to call `SetSessionState()` as the overriding happens at a later time. 

Allow me to look deeper into this `SessionTicketExtension` and make sure skipping it. I am a bit confused on why are we calling `SetSessionState` outside the if... i.e., what's the benefit of calling `uconn.SetSessionState(nil)`?

[the remark this answers is not in the document]
Comment (zeeker999) [reply examples/tls-psk/main.go:93]: ...
> Allow me to look deeper into this `SessionTicketExtension` and make sure skipping it. I am a bit confused on why are we calling `SetSessionState` outside the if... i.e., what's the benefit of calling `uconn.SetSessionState(nil)`?

sets the uconn.HandshakeState.Hello.TicketSupported to true.

[the remark this answers is not in the document]
Comment (gaukas) [reply examples/tls-psk/main.go:93]: Sounds like a terrible idea to call an entire function to set one internal state 🥶. I will allow it I guess...

[the remark this answers is not in the document]
Comment (gaukas) [reply examples/tls-psk/main.go:93]: Should we set `uconn.HandshakeState.Hello.TicketSupported`/`hello.ticketSupported` to true even when we are not using session ticket extension based resumption? I am skeptical. 

Anyways, seemingly just breaking from the switch works, it should be fine for us to skip calling `SetSessionState(cs)` here if the session is for TLS 1.3.

**Comment by gaukas** [comment]
CI build failing. Fixing... 😞

**Pull request merged**